test/e2e/ownerreference_helpers.go

```diff
@@ -126,6 +131,32 @@ var (
 			return framework.HasExactOwners(owners)
 		},
 	}
+
+	// NutanixInfraOwnerReferenceAssertions maps Nutanix Infrastructure types to functions which return an error
+	// if the passed OwnerReferences aren't as expected.
+	NutanixInfraOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+		nutanixMachineKind: func(owners []metav1.OwnerReference) error {
+			// The NutanixMachine must be owned and controlled by a Machine.
+			return framework.HasExactOwners(owners, machineController)
+		},
+		nutanixMachineTemplateKind: func(owners []metav1.OwnerReference) error {
+			// Base NutanixMachineTemplates referenced in a ClusterClass must be owned by the ClusterClass.
+			// NutanixMachineTemplates created for specific Clusters in the Topology controller must be owned by a Cluster.
+			return framework.HasOneOfExactOwners(
+				owners,
+				[]metav1.OwnerReference{clusterOwner},
+				[]metav1.OwnerReference{clusterClassOwner},
+			)
+		},
+		nutanixClusterKind: func(owners []metav1.OwnerReference) error {
+			// NutanixCluster must be owned and controlled by a Cluster.
+			return framework.HasExactOwners(owners, clusterController)
+		},
+		nutanixClusterTemplateKind: func(owners []metav1.OwnerReference) error {
+			// NutanixClusterTemplate must be owned by a ClusterClass.
+			return framework.HasExactOwners(owners, clusterClassOwner)
+		},
```

Comment:
With existing code i faced following error 
<pre>
      [
          <*errors.withStack | 0xc001e1a3a8>{
              error: <*errors.withMessage | 0xc0007bfbe0>{
                  cause: <*errors.errorString | 0xc000b67da0>{
                      s: "wanted [infrastructure.cluster.x-k8s.io/v1beta1/NutanixCluster/false], actual []",
                  },
                  msg: "Unexpected ownerReferences for Secret/quick-start-d0c0iu-registry-creds",
              },
              stack: [0xc6cedfb, 0xaccada6, 0xacc9eb9, 0xb6aa69f, 0xb6ab76c, 0xb6a8dc6, 0xc6cea12, 0xc79b0af, 0xc78f723, 0xc7219fa, 0xc7202f1, 0xc78f688, 0xb686eb3, 0xb69af4d, 0xac94f61],
          },
          <*errors.withStack | 0xc001e1a540>{
              error: <*errors.withMessage | 0xc0007bfc00>{
                  cause: <*errors.errorString | 0xc000390260>{
                      s: "wanted [infrastructure.cluster.x-k8s.io/v1beta1/NutanixCluster/false], actual []",
                  },
                  msg: "Unexpected ownerReferences for Secret/quick-start-d0c0iu-dockerhub-credentials",
              },
              stack: [0xc6cedfb, 0xaccada6, 0xacc9eb9, 0xb6aa69f, 0xb6ab76c, 0xb6a8dc6, 0xc6cea12, 0xc79b0af, 0xc78f723, 0xc7219fa, 0xc7202f1, 0xc78f688, 0xb686eb3, 0xb69af4d, 0xac94f61],
          },
          <*errors.withStack | 0xc001e1a5d0>{
              error: <*errors.withMessage | 0xc0007bfc40>{
                  cause: <*errors.errorString | 0xc0003905a0>{
                      s: "wanted [infrastructure.cluster.x-k8s.io/v1beta1/NutanixCluster/false], actual [controlplane.cluster.x-k8s.io/v1beta1/KubeadmControlPlane/true]",
                  },
                  msg: "Unexpected ownerReferences for Secret/quick-start-d0c0iu-etcd",
              },
              stack: [0xc6cedfb, 0xaccada6, 0xacc9eb9, 0xb6aa69f, 0xb6ab76c, 0xb6a8dc6, 0xc6cea12, 0xc79b0af, 0xc78f723, 0xc7219fa, 0xc7202f1, 0xc78f688, 0xb686eb3, 0xb69af4d, 0xac94f61],
          },
          <*errors.withStack | 0xc001e1a660>{
              error: <*errors.withMessage | 0xc0007bfc60>{
                  cause: <*errors.errorString | 0xc000390780>{
                      s: "wanted [infrastructure.cluster.x-k8s.io/v1beta1/NutanixCluster/false], actual [controlplane.cluster.x-k8s.io/v1beta1/KubeadmControlPlane/true]",
                  },
                  msg: "Unexpected ownerReferences for Secret/quick-start-d0c0iu-proxy",
              },
              stack: [0xc6cedfb, 0xaccada6, 0xacc9eb9, 0xb6aa69f, 0xb6ab76c, 0xb6a8dc6, 0xc6cea12, 0xc79b0af, 0xc78f723, 0xc7219fa, 0xc7202f1, 0xc78f688, 0xb686eb3, 0xb69af4d, 0xac94f61],
          },
          <*errors.withStack | 0xc001e1a8d0>{
              error: <*errors.withMessage | 0xc0007bfc80>{
                  cause: <*errors.errorString | 0xc000391370>{
                      s: "wanted [infrastructure.cluster.x-k8s.io/v1beta1/NutanixCluster/false], actual [bootstrap.cluster.x-k8s.io/v1beta1/KubeadmConfig/true]",
                  },
                  msg: "Unexpected ownerReferences for Secret/quick-start-d0c0iu-bvtbr-lw8zm",
              },
              stack: [0xc6cedfb, 0xaccada6, 0xacc9eb9, 0xb6aa69f, 0xb6ab76c, 0xb6a8dc6, 0xc6cea12, 0xc79b0af, 0xc78f723, 0xc7219fa, 0xc7202f1, 0xc78f688, 0xb686eb3, 0xb69af4d, 0xac94f61],
          },
          <*errors.withStack | 0xc001e1a978>{
              error: <*errors.withMessage | 0xc0007bfca0>{
                  cause: <*errors.errorString | 0xc0003915d0>{
                      s: "wanted [infrastructure.cluster.x-k8s.io/v1beta1/NutanixCluster/false], actual [controlplane.cluster.x-k8s.io/v1beta1/KubeadmControlPlane/true]",
                  },
                  msg: "Unexpected ownerReferences for Secret/quick-start-d0c0iu-ca",
              },
              stack: [0xc6cedfb, 0xaccada6, 0xacc9eb9, 0xb6aa69f, 0xb6ab76c, 0xb6a8dc6, 0xc6cea12, 0xc79b0af, 0xc78f723, 0xc7219fa, 0xc7202f1, 0xc78f688, 0xb686eb3, 0xb69af4d, 0xac94f61],
          },
          <*errors.withStack | 0xc001e1aa08>{
              error: <*errors.withMessage | 0xc0007bfcc0>{
                  cause: <*errors.errorString | 0xc000391770>{
                      s: "wanted [infrastructure.c...

  Gomega truncated this representation as it exceeds 'format.MaxLength'.
  Consider having the object provide a custom 'GomegaStringer' representation
  or adjust the parameters in Gomega's 'format' package.

  Learn more here: https://onsi.github.io/gomega/#adjusting-output
  
  In [It] at: /Users/deepak.muley/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.6.3/framework/ownerreference_helpers.go:133 @ 04/19/24 13:42:36.344

  Full Stack Trace
</pre>

so i added following to check
<pre>
		secretKind: func(owners []metav1.OwnerReference) error {
			// PC Creds, CSI secret must be owned by a NutanixCluster.
			return framework.HasExactOwners(owners, nutanixClusterOwner)
		},
</pre>

also tried following 
<pre>
		secretKind: func(owners []metav1.OwnerReference) error {
			// PC Creds, CSI secret must be owned by a NutanixCluster.
			return framework.HasOneOfExactOwners(
				owners,
				[]metav1.OwnerReference{nutanixClusterOwner},
			)
		},
</pre>

but still fails. trying to find a better way to allow ownerRef to specific secrets instead of blanket all secrets

Reply:
here is the sample yaml of pc-creds secret 
<pre>
  kubectl get secret nutanix-quick-start-helm-addon-cilium-pc-creds -o yaml
apiVersion: v1
data:
  credentials: 
kind: Secret
metadata:
  annotations:
  creationTimestamp: "2024-04-19T19:59:40Z"
  finalizers:
  - nutanixcluster/infrastructure.cluster.x-k8s.io
  labels:
    cluster.x-k8s.io/provider: nutanix
  name: nutanix-quick-start-helm-addon-cilium-pc-creds
  namespace: default
  ownerReferences: <===============================
  - apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: NutanixCluster
    name: nutanix-quick-start-helm-addon-cilium-q76nt
    uid: 5b295fde-8fc6-49a7-a7c9-97db4563b308
  resourceVersion: "1629"
  uid: 60f2f76f-0ad2-4c38-836e-e85cc09dd4a0
type: Opaque
</pre>

but this ref does not exist for dockerhub and registry creds. but only for pc-creds-for-csi and pc-creds

Reply:
i could go ahead with following patch for timebeing 
<pre>
+               secretKind: func(owners []metav1.OwnerReference) error {
+                       // TODO:deepakm-ntnx PC Creds must be owned by a NutanixCluster.
+                       // TODO:deepakm-ntnx CSI secret currently does not have owner.
+                       return nil
+               },
</pre>

but now it fails with following
<pre>
 [FAILED] Timed out after 300.001s.
  Expected success, but got an error:
      <errors.aggregate | len:1, cap:1>: 
      Unexpected ownerReferences for HelmChartProxy/nutanix-csi-snapshot-quick-start-ouzmw7: wanted [cluster.x-k8s.io/v1beta1/Cluster/false], actual []
      [
          <*errors.withStack | 0xc001239518>{
              error: <*errors.withMessage | 0xc0007267c0>{
                  cause: <*errors.errorString | 0xc001299700>{
                      s: "wanted [cluster.x-k8s.io/v1beta1/Cluster/false], actual []",
                  },
                  msg: "Unexpected ownerReferences for HelmChartProxy/nutanix-csi-snapshot-quick-start-ouzmw7",
              },
              stack: [0x12058dfb, 0x10654da6, 0x10653eb9, 0x1103469f, 0x1103576c, 0x11032dc6, 0x12058a12, 0x1212502f, 0x12119723, 0x120ab9fa, 0x120aa2f1, 0x12119688, 0x11010eb3, 0x11024f4d, 0x1061ef61],
          },
      ]
  In [It] at: /Users/deepak.muley/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.6.3/framework/ownerreference_helpers.go:133 @ 04/22/24 14:02:01.494


</pre>

Reply:
have filed respective trackers as specifically secrets needs a different way to handle them